pkg/transparentproxy/validate/validator.go

```diff
@@ -0,0 +1,191 @@
+	clientConnectIP, _ := netip.AddrFromSlice([]byte{127, 0, 0, 6})
+
+	if useIpv6 {
+		serverListenIP, _ = netip.AddrFromSlice([]byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1})
+		clientConnectIP, _ = netip.AddrFromSlice([]byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 6})
+	}
+
+	return &Validator{
+		Config: &Config{
+			ServerListenIP:      serverListenIP,
+			ServerListenPort:    port,
+			ClientConnectIP:     clientConnectIP,
+			ClientRetryInterval: validationInterval,
+		},
+		Logger: logger,
+	}
+}
+
+func (validator *Validator) Run() error {
+	validator.Logger.Info("Starting iptables validation...")
+	sExit := make(chan struct{})
+
+	sError := validator.runServer(sExit)
+	select {
+	case serverErr := <-sError:
+		if serverErr == nil {
+			serverErr = fmt.Errorf("server exited unexpectedly")
+		}
+		validator.Logger.Error(serverErr, "Validation failed")
+		return serverErr
+	default:
+	}
+
+	clientErr := validator.runClient()
+	if clientErr != nil {
+		validator.Logger.Error(clientErr, "Validation failed, client failed to connect to the verification server")
+		close(sExit)
+		return clientErr
+	} else {
+		close(sExit)
+		validator.Logger.Info("Validation passed, iptables rules applied correctly")
+		return nil
+	}
+}
+
+func (validator *Validator) runServer(sExit chan struct{}) chan error {
+	s := LocalServer{
+		logger: validator.Logger,
+		config: validator.Config,
+	}
+
+	sReady := make(chan struct{}, 1)
+	sError := make(chan error, 1)
+	go func() {
+		sError <- s.Run(sReady, sExit)
+	}()
+
+	<-sReady
+	return sError
+}
+
+type LocalServer struct {
+	logger logr.Logger
+	config *Config
+}
+
+func (s *LocalServer) Run(readiness chan struct{}, exit chan struct{}) error {
+	addr := net.JoinHostPort(s.config.ServerListenIP.String(), fmt.Sprintf("%d", s.config.ServerListenPort))
+	s.logger.Info(fmt.Sprintf("Listening on %v", addr))
+
+	config := &net.ListenConfig{}
+	l, err := config.Listen(context.Background(), "tcp", addr)
+	if err != nil {
+		s.logger.Error(err, fmt.Sprintf("error listening on %v", s.config.ServerListenIP))
+		return err
+	}
+
+	go s.handleTcpConnections(l, exit)
+
+	readiness <- struct{}{}
+	<-exit
+	l.Close()
+	return nil
+}
+
+func (s *LocalServer) handleTcpConnections(l net.Listener, cExit chan struct{}) {
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			s.logger.Error(err, "Listener failed to accept connection")
+			return
+		}
+
+		s.logger.Error(err, "Server: a connection has been established")
+		_, _ = conn.Write([]byte(s.config.ServerListenIP.String()))
+		_ = conn.Close()
+
+		select {
+		case <-cExit:
+			return
+		default:
+		}
+	}
+}
+
+func (validator *Validator) runClient() error {
+	c := LocalClient{ServerIP: validator.Config.ClientConnectIP, ServerPort: validator.Config.ClientConnectPort}
+	backoff := retry.WithMaxRetries(validationRetries, retry.NewConstant(validator.Config.ClientRetryInterval))
+	return retry.Do(context.TODO(), backoff, func(ctx context.Context) error {
```

Comment:
missing context

Reply:
The whole validator should be running relatively quickly (at most 10s), is it still needed to make it cancelable? Or do you think a context can be beneficial for other reasons?

Reply:
so I guess we can use `context.Background()` right? I know they are eventually the same but `TODO` indicates that PR is not finished yet